[GTK][WPE][Skia] Visiting wpewebkit.org with Skia/CPU rendering fires assertion

nikolaszimmermann opened the pull request:
#### 1021b2c0a05a9e191f3faad0d531952817579046
<pre>
[GTK][WPE][Skia] Visiting wpewebkit.org with Skia/CPU rendering fires assertion
<a href="https://bugs.webkit.org/show_bug.cgi?id=285768">https://bugs.webkit.org/show_bug.cgi?id=285768</a>

Reviewed by Adrian Perez de Castro.

drawFilteredImageBuffer() tries to determine if it needs to copy the
ImageBuffer backend of the accelerated image, in case of cross-thread
usage (e.g. re-wrapping an existing GPU resource in another thread).
This erroneously ran also for unaccelerated rendering, firing an
assertion that the SkiaGLContext is not available, which is correct. We
shouldn&apos;t attempt to run the copy-ImageBuffer-backend logic during
unaccelerated rendering -- fix that.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer):

Canonical link: <a href="https://commits.webkit.org/288778@main">https://commits.webkit.org/288778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b624b2af195820608bd2c0225a16e1a6d402c9ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38564 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89336 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35269 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11856 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65531 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23380 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2975 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76559 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45821 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2925 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30789 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34318 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73828 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90718 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11527 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8367 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73986 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73188 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18131 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17505 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15946 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2892 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11479 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16955 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11328 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14804 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -330,21 +330,29 @@ void GraphicsContextSkia::drawNativeImageInternal(NativeImage& nativeImage, cons
-        // The 'image' was produced on another thread -- to use it here, we need to create a new ImageBuffer, that wraps the existing GPU resource.
-        auto newSourceImage = sourceImage->copyAcceleratedImageBufferBorrowingBackendRenderTarget();
-        GraphicsContext::drawFilteredImageBuffer(newSourceImage.get(), sourceImageRect, filter, results);
+    auto needsAcceleratedImageBufferCopy = [&]() -> bool {
+        if (m_renderingMode == RenderingMode::Unaccelerated)
+            return false;
+        if (!sourceImage || !sourceImage->skiaGrContext())
+            return false;
+        if (sourceImage->skiaGrContext() == PlatformDisplay::sharedDisplay().skiaGrContext())
+            return false;
+        return true;
+    };
+
+    if (!needsAcceleratedImageBufferCopy()) {
```
I would have calculated the boolean flag directly without using a lambda, but no strong feeling about it, so no need to apply the suggestion without landing if you feel the code is easier to read as-is.

```suggestion
    const bool needsAcceleratedImageBufferCopy =
        m_renderingMode != RenderingMode::Unaccelerated &&
        sourceImage && sourceImage->skiaGrContext()) &&
        sourceImage->skiaGrContext() != PlatformDisplay::sharedDisplay().skiaGrContext());

    if (!needsAcceleratedImageBufferCopy) {
```

nikolaszimmermann replied:
I actually prefer the lambda due to the early returns :-)

aperezdc reviewed (approved):
Patch LGTM. I have a small suggestion that you may optionally apply before landing&mdash;definitely not a blocker, though.

nikolaszimmermann committed:
[GTK][WPE][Skia] Visiting wpewebkit.org with Skia/CPU rendering fires assertion
https://bugs.webkit.org/show_bug.cgi?id=285768

Reviewed by Adrian Perez de Castro.

drawFilteredImageBuffer() tries to determine if it needs to copy the
ImageBuffer backend of the accelerated image, in case of cross-thread
usage (e.g. re-wrapping an existing GPU resource in another thread).
This erroneously ran also for unaccelerated rendering, firing an
assertion that the SkiaGLContext is not available, which is correct. We
shouldn't attempt to run the copy-ImageBuffer-backend logic during
unaccelerated rendering -- fix that.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer):

Canonical link: https://commits.webkit.org/288778@main
